### PR TITLE
configuration_manual: Update aws-s3 and other s3-compatible storages

### PR DESCRIPTION
**Pull request by vmo69**
Introduce the aws-s3 scheme and add more structure.

### CODE REVIEW

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 1]
```diff
@@ -4,6 +4,13 @@
 Amazon S3
 ================
 
+This document covers the Amazon Web Services S3 (Simple Storage Service). In case
+you want to use non AWS S3 storage services, please refer to
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: s/non AWS/another provider's/ ?

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 2]
```diff
@@ -35,13 +42,22 @@ partitions = this would theoretically provide you up to: ``~896,000
...
   plugin {
-    obox_fs = fscache:1G:/var/cache/mails:compress:gz:6:s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
-    obox_index_fs = compress:gz:6:s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
+    obox_fs = fscache:1G:/var/cache/mails:compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
+    obox_index_fs = compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
   }
 
+.. Note::
+  The scheme ``aws-s3`` was newly introduced if your version does not support
+  it yet stick to ``s3``
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: In general I don't like text like this because it quickly becomes out of date.  For example, in 6 months "newly" will feel a bit odd.  In a year, it'll be outright wrong :)

How about:

If the version you are using does not support the ``aws-s3`` scheme, try using the more generic ``s3`` scheme.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 3]
```diff
@@ -60,10 +76,10 @@ by adding the bucket region parameter to the S3 URL:
 .. code-block:: none
 
   plugin {
-    obox_index_fs = compress:gz:6:s3: https://ACCESSKEY:SECRET@host/?bucket=BUCKETNAME&region=eu-central-1
+    obox_index_fs = compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@host/?bucket=BUCKETNAME&region=eu-central-1
   }
 
-The s3 driver also supports the bulk-deletion. The bulk-delete is enabled by default to delete up to 1000 keys with one request. To change this behaviour refer to ``bulk_delete_limit`` at :ref:`http_based_object_storages`
+The ``aws-s3`` and ``s3`` driver also support bulk-deletion. The bulk-delete is enabled by default to delete up to 1000 keys with one request. To change this behaviour refer to ``bulk_delete_limit`` at :ref:`http_based_object_storages`. Bulk delete can only efficiently run on multiple objects if you allow it to, via setting ``obox_max_parallel_deletes`` greater one (refer to :ref:`obox_settings`).
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: s/driver also support/drivers support/
s/bulk-delete/``bulk-delete`` option/
s/you allow it to/configured to do so/

File: source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst [hunk 4]
```diff
@@ -19,6 +23,7 @@ given prefix. Many S3 storages either don't implement the listing at all, or
 it's only used for disaster recovery type of purposes to list all objects. If
 this is the case, you can still use the storage together with Cassandra.
 
+
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst]: Is this addition intentional?

File: source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst [hunk 5]
```diff
@@ -4,8 +4,12 @@
 Other S3-compatible Storages
 ==================================
 
-The configuration is likely to be similar to Amazon S3, but often the bucket
-name needs to be specified explicitly via the bucket parameter:
+This document covers the S3 compatible storage services. In case you want to
+use AWS S3 please refer to :ref:`amazon_s3`.
+
+The configuration for ``s3`` is likely to be similar to Amazon S3 ``aws-s3``,
+but often the bucket name needs to be specified explicitly via the bucket
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst]: s/often the bucket name needs to/generally the bucket name must/

File: source/configuration_manual/mail_location/obox/amazon_s3.rst (hunk 1 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: yes good point thanks.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst (hunk 3 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: thanks

File: source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst (hunk 4 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst]: nope that was a accident.

File: source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst (hunk 5 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/other_s3-compatible_storages.rst]: yes good point!

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 6]
```diff
@@ -35,13 +42,21 @@ partitions = this would theoretically provide you up to: ``~896,000
...
 
   plugin {
-    obox_fs = fscache:1G:/var/cache/mails:compress:gz:6:s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
-    obox_index_fs = compress:gz:6:s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
+    obox_fs = fscache:1G:/var/cache/mails:compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
+    obox_index_fs = compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
   }
 
+.. Note::
+  If the version you are using does not support the ``aws-s3`` scheme, try using the more generic ``s3`` scheme.
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: I just noticed that some of the lines in this file are line wrapped while others aren't.  I don't know what the preferred format is, but I'd say at least the lines you're touching should be properly wrapped/not-wrapped.

Comment (vmo69) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: I cleaned up a bit so at least the things i changed on this page are wrapped now.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 7]
```diff
@@ -35,13 +42,22 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+Configuration
+^^^^^^^^^^^^^
+
 .. code-block:: none
 
   plugin {
-    obox_fs = fscache:1G:/var/cache/mails:compress:gz:6:s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
-    obox_index_fs = compress:gz:6:s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
+    obox_fs = fscache:1G:/var/cache/mails:compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
+    obox_index_fs = compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@BUCKETNAME.s3.amazonaws.com/
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: There should't be extra space after "aws-s3:" (shouldn't have been in the original doc either)

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 8]
```diff
@@ -60,10 +76,15 @@ by adding the bucket region parameter to the S3 URL:
 .. code-block:: none
 
   plugin {
-    obox_index_fs = compress:gz:6:s3: https://ACCESSKEY:SECRET@host/?bucket=BUCKETNAME&region=eu-central-1
+    obox_index_fs = compress:gz:6:aws-s3: https://ACCESSKEY:SECRET@host/?bucket=BUCKETNAME&region=eu-central-1
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: here also

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 9]
```diff
@@ -60,10 +76,15 @@ by adding the bucket region parameter to the S3 URL:
...
 
-The s3 driver also supports the bulk-deletion. The bulk-delete is enabled by default to delete up to 1000 keys with one request. To change this behaviour refer to ``bulk_delete_limit`` at :ref:`http_based_object_storages`
+The ``aws-s3`` and ``s3`` drivers support bulk-deletion. The ``bulk-delete``
+option is enabled by default to delete up to 1000 keys with one request.
+To change this behaviour refer to ``bulk_delete_limit`` at
+:ref:`http_based_object_storages`. Bulk delete can only efficiently run on
+multiple objects if configured to do so, via setting
+``obox_max_parallel_deletes`` greater one (refer to :ref:`obox_settings`).
 
   .. versionadded:: 2.3.10
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: It should also say how s3 and aws-s3 differ (list the default parameters that it uses for s3).

File: source/configuration_manual/mail_location/obox/http_based_object_storages.rst [hunk 10]
```diff
@@ -35,7 +35,14 @@ The parameters common to all object storages include:
...
+| addhdr=<name>:<value>           |Add the specified header to all HTTP requests. This may be useful for load balancing.                                          | none         |
+|                                 |                                                                                                                               |              |
+|                                 |.. versionadded:: 2.3.10 Can be specified multiple times.                                                                      |              |
++---------------------------------+-------------------------------------------------------------------------------------------------------------------------------+--------------+
+| addhdrvar=<name>:<variable>     |Add the specified header to all HTTP requests and set the value to the expanded variables value.                               | none         |
+|                                 |This may be useful for load balancing.                                                                                         |              |
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/http_based_object_storages.rst]: Remove the "This may be useful for load balancing" .. I think also from addhdr. That's only one very specific use case for CDMI which isn't really even used anymore.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 11]
```diff
@@ -35,20 +42,36 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+        .. versionadded:: 2.3.10
+
 Get ACCESSKEY and SECRET from `www.aws.amazon.com <https://aws.amazon.com/>`_
 ``-> My account -> Security credentials -> Access`` credentials. Create the
 ``BUCKETNAME`` from ``AWS Management Console -> S3 -> Create Bucket``.
 
 If the ``ACCESSKEY`` or ``SECRET`` contains any special characters, they can be
 %hex-encoded.
 
+If the ``aws-3`` scheme is used dovecot defaults to append the following url
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: typo, aws-s3

Comment (sirainen) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: Also, Dovecot upper case :)

Comment (sirainen) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: Also, URL uppercase :)

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 12]
```diff
@@ -35,20 +42,36 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+parameters, refer to :ref:`http_based_object_storages` for details.
+
+.. code-block:: none
+
+  auth_protocol=iam&addhdrvar=x-amz-security-token:%{auth:token}&loghdr=x-amz-request-id&loghdr=x-amz-id-2
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: Oh right, this brings to my mind another point: The aws-s3 currently requires IAM, meaning it only works in AWS's VMs. It would be nice if it was possible to disable IAM by using `aws-s3:http://.../?auth_protocol=` but because this string is now appended it overwrites the auth_protocol.. Maybe it should be prepended after the `?` instead so auth_protocol can be overwritten?

And the documentation should explain about using S3 in AWS VMs vs externally..

Comment (sirainen) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: Although I'm not sure how x-amz-security-token works in that situation. I guess if auth:token expands to empty it could avoid adding the header (not sure if it already does). But the loghdrs are the main nice and easy thing about aws-s3.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 13]
```diff
@@ -35,35 +42,80 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+        .. versionadded:: 2.3.10
+
 Get ACCESSKEY and SECRET from `www.aws.amazon.com <https://aws.amazon.com/>`_
 ``-> My account -> Security credentials -> Access`` credentials. Create the
 ``BUCKETNAME`` from ``AWS Management Console -> S3 -> Create Bucket``.
 
 If the ``ACCESSKEY`` or ``SECRET`` contains any special characters, they can be
 %hex-encoded.
 
+If the ``aws-s3`` scheme is used Dovecot defaults to prepend the following URL
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: Is it really "prepend"?  Or does it append?

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 14]
```diff
@@ -35,35 +42,80 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+
+.. code-block:: none
+
+  auth_protocol=iam&addhdrvar=x-amz-security-token:%{auth:token}&loghdr=x-amz-request-id&loghdr=x-amz-id-2
+
+
+Using ``aws-s3`` without IAM
+""""""""""""""""""""""""""""
+
+In order to benefit of the predefined parameter's while not using IAM, e.g.
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: s/benefit of/benefit from/
s/parameter's/parameters/

Comment (jeffpc) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: s/accessing/when accessing/

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 15]
```diff
@@ -35,35 +42,80 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+accessing Amazon S3 from outside of AWS, the ``auth_protocol`` must be
+overwritten. You can overwrite the ``auth_protocol`` by appending it to the
```
Comment (jeffpc) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: s/overwritten/overridden/ ?

File: source/configuration_manual/mail_location/obox/amazon_s3.rst (hunk 13 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: It's prepend now (it changed), we thereby allow users to override defaults.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst (hunk 14 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: thanks a lot.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst (hunk 15 above)
Comment (vmo69) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: yes changed it.

Comment (jeffpc) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: The following sentence also needs a s/overwrite/override/ :)

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 16]
```diff
@@ -35,35 +42,66 @@ partitions = this would theoretically provide you up to: ``~896,000
...
 %hex-encoded.
 
+If the ``aws-s3`` scheme is used Dovecot defaults to prepend the following URL
+parameters, refer to :ref:`http_based_object_storages` for details. The same
+setup can be achieved by using the ``s3`` scheme and adding the parameters
+manually.
+
+.. code-block:: none
+
+  addhdrvar=x-amz-security-token:%{auth:token}&loghdr=x-amz-request-id&loghdr=x-amz-id-2
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: Needs to mention auth_role setting in some way. That using it enables doing IAM lookups, which then overwrites the access key/secret.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst (hunk 13 above)
Comment (sirainen) [reply source/configuration_manual/mail_location/obox/amazon_s3.rst]: (Reminder: Don't click unresolve conversation before clicking comment, or it will lose the comment.)

I wonder where Markus's comment about this went that I see in my email:

> It's prepend now (it changed), we thereby allow users to override defaults.

Anyway now that we removed auth_protocol setting (-> remove it from these changes), there's no longer any difference between prepend and append since the "addhdrvar" and "loghdr" can't be overridden. But I suppose this could still be kept as "prepended"..

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 17]
```diff
@@ -35,35 +42,100 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+instance, otherwise the IMDS will not be reachable. Additionally an IAM role
+must be configured which allows trusted entities, EC2 in this case, to
+assume that role. The role (for example ``s3access``) that will be assumed must
+have the ``AmazonS3FullAccess`` policy attached.
+
+The ``auth_role`` can be configured as a URL parameter it specifies the IAM role
+to be assumed. If no ``auth_role`` is configured, no IAM lookup will be done.
+The IAM requests are targeted by default to the IMDSv2 service in EC2
+(``169.254.169.254``). It is possible to override this default by appending
+``auth_host=NEWHOST`` to the URL.
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: I guess could mention also auth_port. And say that those are mainly useful for testing.

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 18]
```diff
@@ -35,35 +42,100 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+Using IAM allows running Dovecot with S3 Storage while not keeping the
+credentials in the configuration.
+
+A requirement for using IMDSv2 is that Dovecot is running on an AWS EC2
+instance, otherwise the IMDS will not be reachable. Additionally an IAM role
+must be configured which allows trusted entities, EC2 in this case, to
+assume that role. The role (for example ``s3access``) that will be assumed must
+have the ``AmazonS3FullAccess`` policy attached.
+
+The ``auth_role`` can be configured as a URL parameter it specifies the IAM role
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: s/it specifies/which specifies/

File: source/configuration_manual/mail_location/obox/amazon_s3.rst [hunk 19]
```diff
@@ -35,35 +42,101 @@ partitions = this would theoretically provide you up to: ``~896,000
...
+must be configured which allows trusted entities, EC2 in this case, to
+assume that role. The role (for example ``s3access``) that will be assumed must
+have the ``AmazonS3FullAccess`` policy attached.
+
+The ``auth_role`` can be configured as a URL parameter which specifies the IAM
+role to be assumed. If no ``auth_role`` is configured, no IAM lookup will be
+done. The IAM requests are targeted by default to the IMDSv2 service in EC2
+(``169.254.169.25:80``). It is possible to override this default by appending
+``auth_host=NEWHOST:NEWPORT`` to the URL, this is mainly used for testing
+purpose's.
```
Comment (sirainen) [inline source/configuration_manual/mail_location/obox/amazon_s3.rst]: "purposes"
